network/discovery/src/lib.rs

```diff
@@ -68,6 +74,62 @@ pub static NETWORK_KEY_MISMATCH: Lazy<IntGaugeVec> = Lazy::new(|| {
     .unwrap()
 });
 
+/// A generic listener for network peer discovery
+pub struct DiscoveryChangeListener {
+    discovery_source: DiscoverySource,
+    network_context: Arc<NetworkContext>,
+    update_channel: channel::Sender<ConnectivityRequest>,
+    source_stream: DiscoveryChangeStream,
+}
+
+enum DiscoveryChangeStream {}
+
+impl Stream for DiscoveryChangeStream {
+    type Item = PeerSet;
+
+    fn poll_next(self: Pin<&mut Self>, _cx: &mut Context<'_>) -> Poll<Option<Self::Item>> {
+        todo!()
+    }
+}
+
+impl DiscoveryChangeListener {
+    pub fn start(self, executor: &Handle) {
+        executor.spawn(Box::pin(self).run());
+    }
+
+    async fn run(mut self: Pin<Box<Self>>) {
+        info!(
+            NetworkSchema::new(&self.network_context),
+            "{} Starting {} Discovery", self.network_context, self.discovery_source
+        );
+
+        while let Some(update) = self.source_stream.next().await {
```

Comment:
In your File based discovery there is a chance that if the file loaded errors or something, then it will return None, if that happens then this will exit this loop causing the discovery mechanism to cease working till the node restarts. So you either need to fix this to properly handle the sources returning None, or fix the File discovery to never return None.

Reply:
Okay, will address this in file based discovery.